catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/SourcePoller.java

```diff
@@ -13,110 +13,285 @@
+    if (!sourceKeysToRemoveFromTheCache.isEmpty()) {
+      if (LOGGER.isDebugEnabled()) {
+        LOGGER.debug(
+            "Found {} entries to remove from the cache: {}. This is caused by when sources are destroyed or modified in a way that causes their SourceKey to be different since the last poll",
+            sourceKeysToRemoveFromTheCache.size(),
+            StringUtils.join(sourceKeysToRemoveFromTheCache, ", "));
+      }
+      sourceKeysToRemoveFromTheCache
+          .stream()
+          .peek(
+              sourceKey1 -> LOGGER.debug("Removing entry sourceKey {} from the cache", sourceKey1))
+          .forEach(cache::invalidate);
+    } else {
+      LOGGER.trace(
+          "Found no entries to remove from the cache. There has been no sources destroyed or modified in a way that causes their SourceKey to be different since the last poll");
+    }
 
-    runner.shutdown();
+    final int currentSourcesCount = currentSources.size();
 
-    handle.cancel(true);
+    // Loop through each of the currentSources
+    LOGGER.trace(
+        "Starting a process to get the value for each of the {} current sources",
+        currentSourcesCount);
+    final CompletionService<Runnable> completionService =
+        new ExecutorCompletionService<>(pollTimeoutWatcherThreadPool);
+    for (Map.Entry<SourceKey, Source> sourceKeySourceEntry :
+        currentSourceKeyAndSourceMap.entrySet()) {
+      final SourceKey sourceKey = sourceKeySourceEntry.getKey();
+      final Source source = sourceKeySourceEntry.getValue();
 
-    scheduler.shutdownNow();
+      final Future<T> future;
+      try {
+        future = pollThreadPool.submit(() -> getValue(source));
+      } catch (RejectedExecutionException e) {
+        // TODO
+        throw new IllegalStateException("TODO");
+      }
+
+      final Callable<Runnable> callable =
+          () -> {
+            try {
+              final T newValue = future.get(timeout, timeoutTimeUnit);
+              if (newValue == null) {
+                throw new IllegalArgumentException("getValueFunction may not return null");
+              }
+
+              return () -> cacheNewValue(sourceKey, newValue);
+            } catch (TimeoutException e) {
+              return () -> handleTimeout(sourceKey);
+            } catch (ExecutionException e) {
+              final Throwable cause = e.getCause();
+              if (cause instanceof RuntimeException) {
+                return () -> handleException(sourceKey, (RuntimeException) cause);
+              } else {
+                // TODO
+                throw new IllegalStateException("TODO");
+              }
+            } catch (InterruptedException e) {
+              // TODO
+              throw new IllegalStateException("TODO");
+            }
+          };
+
+      try {
+        completionService.submit(callable);
+      } catch (RejectedExecutionException e) {
+        // TODO
+        throw new IllegalStateException("TODO");
```

Comment:
❓When we add the code to this catch block, we should consider adding some retry logic, or log something obvious so the admin knows a source is not being polled.